### DISCUSSION TITLE
Libvips vips_image_new_from_file() some formats, file handle not closed in timely manner.

### DISCUSSION THREAD

**Discussion by emerth** [libvips/libvips#3185] [Q&A]
Some file formats reads via vips_image_new_from_file() do not close file handle in a timely manner. 

Q: Can I force VIPS to close a file opened with vips_image_new_from_file()?

Formats affected: GIF, PGM, PNM, PPM. 

Work flow, sans error checking:

```
    unsigned char* load_an_image(const char *filename) {
        unsigned char *my_pixels = NULL;
        VipsImage* in = vips_image_new_from_file(filename, VIPS_ACCESS_RANDOM);

        width = vips_image_get_width(in);
        height = vips_image_get_height(in);

        boolean interp_OK = (vips_image_get_interpretation(in) == VIPS_INTERPRETATION_sRGB);
        boolean bytes_per_band_OK = (vips_format_sizeof(fmt) == 1);
        boolean format_uchar_ok = (fmt == VIPS_FORMAT_UCHAR);
        boolean format_char_ok = (fmt == VIPS_FORMAT_CHAR);
        verify_everything_is_ok(interp_OK, bytes_per_band_OK, format_uchar_ok, format_char_ok);

        unsigned char* vips_pixels = (unsigned char*)vips_image_get_data(in);
        
        my_pixels = new unsigned char[width * height * 4];
                
        copy_vips_pixels_to_my_pixels(vips_pixels, my_pixels, width, height);
        
        g_object_unref(in);

        return my_pixels; // my_pixels to be processed by unrelated non-VIPS code.
    }

  // After a call to the above function: for GIF, PGM, PNM, PPM files the file handle is still open for another 60 to 90 seconds.
  // Other file formats: the file handle is closed by now.
```

My specific issue is that the processing  done on a file's pixels might result in a decision to delete the file. If the file handle is still held open by VIPS then I cannot delete the file. 

(I use libvips to load small to medium size images entire into RAM for processing elsewhere. My use case is that VIPS can read a large variety of image formats, not that VIPS can do so in a memory efficient manner: my images are not large. So I am using vips_image_get_data() to slurp the entire file. This is by design.)

**Comment by jcupitt** [comment]
Hi @emerth,

These format loaders all use `MapViewOfFile()` to map the source file into memory:

https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-mapviewoffile

libvips is closing the file handle, but behind the scenes, windows is keeping it open as part of the mapping process.

A possible workaround is to load the source image into memory, then use `vips_image_new_from_buffer()`. I would only do this for these formats though -- many formals will be much quicker if you let libvips do the open.

Did you see `vips_image_write_to_memory()`? You can do eg.:

```C
VipsImage *image = vips_image_new_from_file(..);
size_t image_size;
void *my_memory_area = vips_image_write_to_memory( image, &image_size );
```

Now `my_memory_area` is a large memory area owned by you containing the decompressed pixels, which you must free with `free()`.

**Reply by emerth** [reply]
That's nice, convenient. Thanks!

**Comment by jcupitt** [comment]
The naming is supposed to show whether the memory area contains a simple array of uncompressed pixel data (`new_from_memory`, `write_to_memory`), or an image in a compressed representation, like jpeg (`new_from_buffer`, `write_to_buffer`).

**Comment by kleisauke** [comment]
Was this tested on Windows? If so, I think the file is being held open in libvips' cache. Please see the documentation of  [`vips_cache_set_max_files()`](https://www.libvips.org/API/current/VipsOperation.html#vips-cache-set-max-files) for a way to disable this behavior.

(AFAIK, this problem only occurs on Windows, other platforms let you delete open files)

**Reply by jcupitt** [reply]
Oh, true, it could be the cache.

**Reply by kleisauke** [reply]
Perhaps we could consider making the file cache opt-in on Windows? This seems to be a common issue for Windows users, see for example issue https://github.com/kleisauke/net-vips/issues/6 and the linked issues to it.

We might also consider applying https://src.fedoraproject.org/rpms/mingw-glib2/blob/rawhide/f/0001-Use-CreateFile-on-Win32-to-make-sure-g_unlink-always.patch to GLib, so that open files can be deleted on Windows (i.e. it opens files with the `FILE_SHARE_DELETE` flag by default).

**Reply by jcupitt** [reply]
Patching glib feels bad. How about pasting most of `g_file_open()` into `vips__open()`? File opening is a really crucial operation, so it makes sense that we'd want to have very deep control over all the details.

**Reply by kleisauke** [reply]
Indeed, though I don't like pasting that into `vips__open()` either (which would mean that we need to do `g_utf8_to_utf16()` by ourselves - https://github.com/libvips/libvips/commit/d48f80130b8fd13192f46104ffe76bf7641d7e9f).

It's a bit unfortunate that Windows does not have an equivalent `oflag` of `FILE_SHARE_DELETE` for the `_open`/`_wopen` calls, you are forced to use the `CreateFile*` APIs for this. :(

The good thing about this "do not delete if file is open" behavior is that it caught issues in the past (e.g. #1370). I suppose you can also call [`vips_image_invalidate_all()`](https://www.libvips.org/API/current/VipsImage.html#vips-image-invalidate-all) on specific images before deletion, so there is no need to disable the entire file cache.

**Reply by emerth** [reply]
Some more info: calling vips_cache_set_max_files(0) at the start of the program resolves the issue with PGM, PNM and PPM. It does not resolve the issue with GIF.

**Reply by kleisauke** [reply]
> It does not resolve the issue with GIF.

I could reproduce this, and I think this revision range would fix that: https://github.com/libvips/libvips/compare/HEAD...kleisauke:nsgifload-remove-seq-flags

Will test this later on Windows (and open a PR if it works), as it is now getting late in UTC+1. :sweat_smile:

**Reply by kleisauke** [reply]
I just opened PR for this #3189. This test script now seems to work for me on Windows:

<details>
  <summary>Details</summary>

```python
import logging
import os
import gc

import pyvips

logging.basicConfig(level=logging.DEBUG)
pyvips.leak_set(True)
pyvips.cache_set_max_files(0)

# https://storage.googleapis.com/downloads.webmproject.org/webp/images/dancing-banana.gif
filename = 'dancing-banana.gif'

# sanity check webpload
# https://storage.googleapis.com/downloads.webmproject.org/webp/images/dancing_banana2.lossless.webp
#filename = 'dancing_banana2.lossless.webp'

try:
    #im = pyvips.Image.new_from_file(filename, access=pyvips.Access.SEQUENTIAL)
    im = pyvips.Image.new_from_file(filename)
    buf = im.write_to_memory()
    im = None # explicitly unset to force unref
except Exception as e:
    print(f'{e}')
finally:
    # had no effect previously
    gc.collect()

    try:
        os.remove(filename)
    except OSError as e:
        # previously this would print:
        # [WinError 5] Access is denied: 'dancing-banana.gif'
        print(f'{filename} could not be deleted - reason {e}')
```
</details>

> I think this revision range would fix that: https://github.com/libvips/libvips/compare/HEAD...kleisauke:nsgifload-remove-seq-flags

FWIW, I'll open another PR for this, it needs to be properly benchmarked first.

**Reply by jcupitt** [reply]
Windows behaviour with MapViewOfFile and file open/close is very opaque :( Well done for tracking this down.
